dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/command/handler/RecoverFailureTaskCommandHandler.java

```diff
@@ -0,0 +1,218 @@
+                                                            final WorkflowExecuteContextBuilder workflowExecuteContextBuilder) {
+        final ProcessInstance workflowInstance = workflowExecuteContextBuilder.getWorkflowInstance();
+        final Map<String, TaskInstance> taskInstanceMap = super.getValidTaskInstance(workflowInstance)
+                .stream()
+                .collect(Collectors.toMap(TaskInstance::getName, Function.identity()));
+
+        final Set<String> needRecreateTasks = taskInstanceMap.values()
+                .stream()
+                .filter(this::isTaskNeedRecreate)
+                .map(TaskInstance::getName)
+                .collect(Collectors.toSet());
+
+        final IWorkflowGraph workflowGraph = workflowExecuteContextBuilder.getWorkflowGraph();
+        final BiConsumer<String, Set<String>> historyTaskInstanceMarker = (task, successors) -> {
+            boolean isTaskNeedRecreate = needRecreateTasks.contains(task) || workflowGraph.getPredecessors(task)
+                    .stream()
+                    .anyMatch(needRecreateTasks::contains);
+            // If the task instance need to be recreated, then will mark the task instance invalid.
+            // and the TaskExecutionRunnable will not contain the task instance.
+            if (isTaskNeedRecreate) {
+                needRecreateTasks.add(task);
+                if (taskInstanceMap.containsKey(task)) {
+                    taskInstanceDao.markTaskInstanceInvalid(Lists.newArrayList(taskInstanceMap.get(task)));
+                    taskInstanceMap.remove(task);
+                }
+            }
+            // If the task instance need to be recovered, then will mark the task instance to submit.
+            // and the TaskExecutionRunnable will contain the task instance and pass the creation step.
+            if (isTaskNeedRecover(taskInstanceMap.get(task))) {
+                final TaskInstance taskInstance = taskInstanceMap.get(task);
+                taskInstance.setState(TaskExecutionStatus.SUBMITTED_SUCCESS);
+                taskInstanceDao.upsertTaskInstance(taskInstance);
+            }
+        };
+
+        final WorkflowGraphBfsVisitor workflowGraphBfsVisitor = WorkflowGraphBfsVisitor.builder()
+                .onWorkflowGraph(workflowGraph)
+                .taskDependType(workflowInstance.getTaskDependType())
+                .fromTask(parseStartNodesFromWorkflowInstance(workflowExecuteContextBuilder))
+                .doVisitFunction(historyTaskInstanceMarker)
+                .build();
+        workflowGraphBfsVisitor.visit();
+        return new ArrayList<>(taskInstanceMap.values());
+    }
+
+    /**
+     * Whether the task need to be recreated.
+     * <p> If the task state is FAILURE and KILL, then will mark the task invalid and recreate the task.
+     */
+    private boolean isTaskNeedRecreate(final TaskInstance taskInstance) {
+        return taskInstance.getState() == TaskExecutionStatus.FAILURE
```

Comment:
Need to check task retry

Reply:
We don't need to check the retry, since we will create a new task instance no matter if the retry times are over the max retry times, since this is repeat running workflow instance.